src/cmdline.rs

```diff
@@ -71,13 +72,48 @@ fn usage() -> Command {
 /// Parse the commandline into a `Command` to execute.
 pub fn parse() -> Command {
     trace!("parse");
-    let matches = get_app().get_matches();
-
+    let cwd = if let Ok(cwd) = env::current_dir() {
+        cwd
+    } else {
+        println!("sccache: Couldn't determine current working directory");
+        return usage();
+    };
     // The internal start server command is passed in the environment.
     let internal_start_server = match env::var("SCCACHE_START_SERVER") {
         Ok(val) => val == "1",
         Err(_) => false,
     };
+    let mut args: Vec<_> = env::args_os().collect();
+    if ! internal_start_server {
+        if let Ok(exe) = env::current_exe() {
+            match exe.file_stem().and_then(|s| s.to_str()).map(|s| s.to_lowercase()) {
+                // If the executable has its standard name, do nothing.
+                Some(ref e) if e == env!("CARGO_PKG_NAME") => {}
+                // Otherwise, if it was copied/hardlinked under a different $name, act
+                // as if it were invoked with `sccache $name`, but avoid $name resolving
+                // to ourselves again if it's in the PATH.
+                _ => {
+                    let path = env::var_os("PATH");
+                    match which_in(exe.file_name().unwrap(), path.as_ref(), &cwd) {
+                        Ok(ref full_path) if full_path.canonicalize().unwrap() == exe.canonicalize().unwrap() => {
+                            if let Some(dir) = full_path.parent() {
+                                let path = env::join_paths(env::split_paths(&path.unwrap()).filter(|p| p != dir)).ok();
+                                match which_in(exe.file_name().unwrap(), path, &cwd) {
```

Comment:
Hum, I thought the `which` crate had exposed an API that gave you an iterator of results, but I guess not. That would be useful here. (You could file that on the upstream which-rs crate if you were so motivated, I rewrote most of it for sccache a while back.)